app/components/members/table_component.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Members
+  class TableComponent < ::TableComponent
+    options :authorize_update, :available_roles, :is_filtered
+    columns :name, :mail, :roles, :groups, :status
+    sortable_columns :name, :mail, :status
+
+    def initial_sort
+      %i[name asc]
+    end
+
+    def headers
+      columns.map do |name|
+        [name.to_s, header_options(name)]
+      end
+    end
+
+    def header_options(name)
+      { caption: User.human_attribute_name(name) }
+    end
+
+    ##
+    # Adjusts the order so that users are joined to support
+    # sorting by their attributes
+    def sort_collection(query, sort_clause, sort_columns)
+      super(join_users(query), sort_clause, sort_columns)
+    end
+
+    def join_users(query)
+      query.joins(:principal).references(:principal)
```

Comment:
[`references`](http://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-references) is used with `includes` to tell ActiveRecord that it needs to perform a join instead of lazy loading the included items later, so we can use it in where clauses:

```ruby
[8] pry(main)> Member.includes(:principal).where("users.login = ?", "admin").to_a
  Member Load (16.0ms)  SELECT "members".* FROM "members" WHERE (users.login = 'admin')
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "users"
LINE 1: SELECT "members".* FROM "members" WHERE (users.login = 'admi...

[9] pry(main)> Member.includes(:principal).references(:principal).where("users.login = ?", "admin").to_a
  SQL (4.1ms)  SELECT "members"."id" AS t0_r0, "members"."user_id" AS t0_r1, "members"."project_id" AS t0_r2, "members"."created_at" AS t0_r3, "members"."updated_at" AS t0_r4, "users"."id" AS t1_r0, "users"."login" AS t1_r1, "users"."firstname" AS t1_r2, "users"."lastname" AS t1_r3, "users"."mail" AS t1_r4, "users"."admin" AS t1_r5, "users"."status" AS t1_r6, "users"."last_login_on" AS t1_r7, "users"."language" AS t1_r8, "users"."auth_source_id" AS t1_r9, "users"."created_at" AS t1_r10, "users"."updated_at" AS t1_r11, "users"."type" AS t1_r12, "users"."identity_url" AS t1_r13, "users"."first_login" AS t1_r14, "users"."force_password_change" AS t1_r15, "users"."failed_login_count" AS t1_r16, "users"."last_failed_login_on" AS t1_r17, "users"."consented_at" AS t1_r18 FROM "members" LEFT OUTER JOIN "users" ON "users"."id" = "members"."user_id" WHERE (users.login = 'admin')
=> [#<Member:0x000000011d71df40 id: 1, user_id: 4, project_id: 1, created_at: Mon, 08 May 2023 10:41:28.527218000 UTC +00:00, updated_at: Mon, 08 May 2023 10:41:28.531846000 UTC +00:00>,
 #<Member:0x000000011d71bc40 id: 2, user_id: 4, project_id: 2, created_at: Mon, 08 May 2023 10:41:29.929903000 UTC +00:00, updated_at: Mon, 08 May 2023 10:41:29.931310000 UTC +00:00>]
```

For `joins` it is not needed and does not change anything:

```ruby
[10] pry(main)> Member.joins(:principal).where("users.login = ?", "admin").to_a
  Member Load (2.4ms)  SELECT "members".* FROM "members" INNER JOIN "users" ON "users"."id" = "members"."user_id" WHERE (users.login = 'admin')
=> [#<Member:0x000000011d7191c0 id: 1, user_id: 4, project_id: 1, created_at: Mon, 08 May 2023 10:41:28.527218000 UTC +00:00, updated_at: Mon, 08 May 2023 10:41:28.531846000 UTC +00:00>,
 #<Member:0x000000011d719120 id: 2, user_id: 4, project_id: 2, created_at: Mon, 08 May 2023 10:41:29.929903000 UTC +00:00, updated_at: Mon, 08 May 2023 10:41:29.931310000 UTC +00:00>]

[11] pry(main)> Member.joins(:principal).references(:principal).where("users.login = ?", "admin").to_a
  Member Load (28.7ms)  SELECT "members".* FROM "members" INNER JOIN "users" ON "users"."id" = "members"."user_id" WHERE (users.login = 'admin')
=> [#<Member:0x000000011d716600 id: 1, user_id: 4, project_id: 1, created_at: Mon, 08 May 2023 10:41:28.527218000 UTC +00:00, updated_at: Mon, 08 May 2023 10:41:28.531846000 UTC +00:00>,
 #<Member:0x000000011d716560 id: 2, user_id: 4, project_id: 2, created_at: Mon, 08 May 2023 10:41:29.929903000 UTC +00:00, updated_at: Mon, 08 May 2023 10:41:29.931310000 UTC +00:00>]
```

so it's a bit redundant, but also does no harm.